plugins/inputs/azure_monitor/azure_monitor.go

```diff
@@ -0,0 +1,856 @@
+	err := am.checkConfigValidation()
+
+	if err != nil {
+		return err
+	}
+
+	err = am.getAccessToken()
+
+	if err != nil {
+		return err
+	}
+
+	err = am.createResourceGroupTargetsFromSubscriptionTargets()
+
+	if err != nil {
+		return err
+	}
+
+	err = am.createResourceTargetsFromResourceGroupTargets()
+
+	if err != nil {
+		return err
+	}
+
+	err = am.getResourceTargetsMetrics()
+
+	if err != nil {
+		return err
+	}
+
+	am.setResourceTargetsAggregation()
+	am.checkTargetsMaximumMetrics()
+
+	am.Log.Info("Total targets: ", len(am.ResourceTargets))
+
+	return nil
+}
+
+func (am *AzureMonitor) checkConfigValidation() error {
+	if am.SubscriptionID == "" {
+		return fmt.Errorf("subscription_id is empty or missing. Please check your configuration")
+	}
+
+	if am.ClientID == "" {
+		return fmt.Errorf("client_id is empty or missing. Please check your configuration")
+	}
+
+	if am.ClientSecret == "" {
+		return fmt.Errorf("client_secret is empty or missing. Please check your configuration")
+	}
+
+	if am.TenantID == "" {
+		return fmt.Errorf("tenant_id is empty or missing. Please check your configuration")
+	}
+
+	if len(am.ResourceTargets) == 0 && len(am.ResourceGroupTargets) == 0 && len(am.SubscriptionTargets) == 0 {
+		return fmt.Errorf("there is no target to collect metrics from. Please check your configuration")
+	}
+
+	for index, resourceTarget := range am.ResourceTargets {
+		if resourceTarget.ResourceID == "" {
+			return fmt.Errorf(
+				"resource target #%d resource_id is empty or missing. Please check your configuration", index+1)
+		}
+	}
+
+	for resourceGroupIndex, resourceGroupTarget := range am.ResourceGroupTargets {
+		if resourceGroupTarget.ResourceGroup == "" {
+			return fmt.Errorf(
+				"resource group target #%d resource_group is empty or missing. Please check your configuration",
+				resourceGroupIndex+1)
+		}
+
+		for resourceIndex, resource := range resourceGroupTarget.Resources {
+			if resource.ResourceType == "" {
+				return fmt.Errorf(
+					"resource group target #%d resource #%d resource_type is empty or missing. Please check your configuration",
+					resourceGroupIndex+1, resourceIndex+1)
+			}
+		}
+	}
+
+	for index, target := range am.SubscriptionTargets {
+		if target.ResourceType == "" {
+			return fmt.Errorf(
+				"subscription target #%d resource_type is empty or missing. Please check your configuration", index+1)
+		}
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) createResourceGroupTargetsFromSubscriptionTargets() error {
+	if len(am.SubscriptionTargets) == 0 {
+		return nil
+	}
+
+	am.Log.Info("Creating resource group targets from subscription targets")
+
+	apiURL := am.buildSubscriptionResourceGroupsAPIURL()
+	body, err := am.getTargetResponseBody(apiURL)
+
+	if err != nil {
+		return err
+	}
+
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return err
+	}
+
+	for _, value := range bodyData["value"].([]interface{}) {
+		resourceGroup := value.(map[string]interface{})["name"].(string)
```

Comment:
This might panic if the `bodyData` is not of the expected format (e.g. due to backend changed format or similar). Please check the type-assertions and gracefully return an error in that case to avoid tearing-down telegraf.